activerecord/lib/active_record/relation/predicate_builder/association_query_value.rb

```diff
@@ -27,12 +27,12 @@ def ids
         end
 
         def primary_key
-          associated_table.join_primary_key
+          Array(associated_table.join_primary_key)
         end
 
         def convert_to_id(value)
-          if value.respond_to?(primary_key)
-            value.public_send(primary_key)
+          if primary_key.all? { |pk| value.respond_to?(pk) }
```

Comment:
I don't particularly like this approach, I think `value.respond_to?(primary_key)` is being too implicit if our intention is to check whether the `value` is an instance of `ActiveRecord::Base`. At the same time checking for `value.respond_to?(primary_key)` is abstract enough to support any objects that respond to pk column names like `BlogPost.where(comments: [OpenStruct.new(blog_id: 1, id: 1]).to_a`
Not sure if we intentionally wanted to support a duck-typing API but I decided to keep it consistent.